lib/internal/streams/pipeline.js

```diff
@@ -54,6 +56,19 @@ function destroyer(stream, reading, writing, final, callback) {
       stream
         .once('end', _destroy)
         .once('error', _destroy);
+    } else if (
+      !err &&
+      (!final || !stream.readable) &&
+      state &&
+      state.autoDestroy &&
+      state.emitClose &&
+      state.closed === false
+    ) {
+      // 'eos' will invoke the callback early on 'finish' and 'end' due to
+      // compat reasons. Try to soft detect whether 'close' will be emitted.
+      stream
+        .once('error', _destroy)
+        .once('close', _destroy);
     } else {
       _destroy(err);
```

Comment:
I would like to suggest a `setImmediate(_destroy, err)` here... in order to give `'close'` a chance to be invoked.